worker/sshserver/server.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshserver
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/juju/errors"
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/juju/state"
+)
+
+// NewSSHServer returns an embedded SSH server. This server does a few things,
+// reuqests that come in must go through the jump server. The jump server will
+// pipe the connection and pass it into an in-memory instance of another SSH server.
+//
+// This second SSH server (seen within directTCPIPHandlerClosure) will handle the
+// the termination of the SSH connections, note, it is not listening on any ports
+// because we are passing the piped connection to it, essentially allowing the following
+// to work (despite only having one server listening):
+// - `ssh -J controller:2223 ubuntu@app.controller.model`
+//
+// TODO(ale8k): Word this comment better later explaining why the host routing works.
+func NewSSHServer(sp *state.StatePool, jumpHostKey string) (*ssh.Server, error) {
+	jumpHostSigner, err := gossh.ParsePrivateKey([]byte(jumpHostKey))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	server := &ssh.Server{
+		Addr: ":2223",
+		PublicKeyHandler: func(ctx ssh.Context, key ssh.PublicKey) bool {
+			return true
+		},
+		PasswordHandler: func(ctx ssh.Context, password string) bool {
+			return true
+		},
+		ChannelHandlers: map[string]ssh.ChannelHandler{
+			"direct-tcpip": directTCPIPHandlerClosure(sp),
+		},
+	}
+
+	server.AddHostKey(jumpHostSigner)
+
+	return server, nil
+}
+
+// directTCPIPHandlerClosure is a closure that returns a direct-tcpip handler passing in state
+// to check the permissions the user has for the model containing this unit.
+func directTCPIPHandlerClosure(sp *state.StatePool) ssh.ChannelHandler {
```

Comment:
Again, the state pool is a bit heavy weight here, use an interface so we can see what the actual scope of the methods are.